base/multidimensional.jl

```diff
@@ -14,19 +14,34 @@ export CartesianIndex, CartesianRange
 linearindexing{A<:BitArray}(::Type{A}) = LinearFast()
 
 # CartesianIndex
-immutable CartesianIndex{N}
+immutable CartesianIndex{N} <: AbstractCartesianIndex{N}
     I::NTuple{N,Int}
     CartesianIndex(index::NTuple{N,Integer}) = new(index)
 end
 
 CartesianIndex{N}(index::NTuple{N,Integer}) = CartesianIndex{N}(index)
-@generated function (::Type{CartesianIndex{N}}){N}(index::Integer...)
+(::Type{CartesianIndex})(index::Integer...) = CartesianIndex(index)
+(::Type{CartesianIndex{N}}){N}(index::Integer...) = CartesianIndex(index)
+# Allow passing tuples smaller than N
+@generated function (::Type{CartesianIndex{N}}){N,M}(index::NTuple{M,Integer})
```

Comment:
Ah, good catch, thank you.  I just made the immediate fix for now, but it really is remarkable how just a few intrinsics allow for more powerful generic code.